File: challenges/exercism/javascript/easy/difference-of-squares.js
/** Representation of an arithmetic series. */
export class Squares {
  constructor(n) {
    this.n = n;
  }

  get sumOfSquares() {
    return (this.n * (this.n + 1) * (2 * this.n + 1)) / 6;
  }

  get squareOfSum() {
    return ((this.n / 2) * (1 + this.n)) ** 2;
  }

  get difference() {
    return this.squareOfSum - this.sumOfSquares;
  }
}

let square5 = new Squares(5);
console.assert(square5.sumOfSquares === 55, 'sumOfSquares');
console.assert(square5.squareOfSum === 225, 'squareOfSum');

File: challenges/exercism/rust/src/easy/difference_of_squares.rs
pub fn square_of_sum(n: u32) -> u32 {
    (0..=n).sum::<u32>().pow(2)

    // fold:
    // (0..=n).fold(0, |acc, n| acc + n).pow(2)

    // formula:
    // ((n * (n + 1)) / 2).pow(2)
}

pub fn sum_of_squares(n: u32) -> u32 {
    (0..=n).map(|n| n.pow(2)).sum()

    // formula:
    // (n * (n + 1) * (2 * n + 1)) / 6
}

pub fn difference(n: u32) -> u32 {
    square_of_sum(n) - sum_of_squares(n)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn square_of_sum_1() {
        assert_eq!(square_of_sum(1), 1);
    }

    #[test]
    fn square_of_sum_5() {
        assert_eq!(square_of_sum(5), 225);
    }

    #[test]
    fn sum_of_squares_1() {
        assert_eq!(sum_of_squares(1), 1);
    }

    #[test]
    fn sum_of_squares_5() {
        assert_eq!(sum_of_squares(5), 55);
    }

    #[test]
    fn difference_100() {
        assert_eq!(difference(100), 25_164_150);
    }
}
